### PR TITLE
Special ic 4507

### PR DESCRIPTION
**Pull request by WilkAndy**
The coder can now designate certain combinations of plasticity models as 'special'. These combinations can then be given special algorithms to determine the set of initially active constraints. This can help markedly in the convergence of the return-map algorithm, just like choosing a good initial condition for a Newton-Raphson procedure.

### CODE REVIEW

**Comment by WilkAndy** [comment]
By the way, a previous PR (#4672) failed because of a problem in moose_test:

utils/random.threads_verification_uo.......................................................... FAILED (CRASH)
Working Directory: /home/moosetest/moose_build/client_0_/moose/test/tests/utils/random
utils/random.threads_verification_uo: Running command: mpiexec -host localhost -n 1 /home/moosetest/moose_build/client_0_/moose/test/moose_test-opt --n-threads=2 -i random_uo.i --error-unused
utils/random.threads_verification_uo: 
utils/random.threads_verification_uo: ===================================================================================
utils/random.threads_verification_uo: =   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
utils/random.threads_verification_uo: =   PID 12021 RUNNING AT localhost
utils/random.threads_verification_uo: =   EXIT CODE: 139
utils/random.threads_verification_uo: =   CLEANING UP REMAINING PROCESSES
utils/random.threads_verification_uo: =   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
utils/random.threads_verification_uo: ===================================================================================
utils/random.threads_verification_uo: YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
utils/random.threads_verification_uo: This typically refers to a problem with your application.
utils/random.threads_verification_uo: Please see the FAQ page for debugging suggestions
utils/random.threads_verification_uo: 
utils/random.threads_verification_uo.......................................................... FAILED (CRASH) (reprint)

**Comment by WilkAndy** [comment]
Hmmm, same problem again.  I don't think this is my fault.  Can you fix it?

a

**Comment by WilkAndy** [comment]
Oops, sorry, i'll fix my silly mistake!

a

**Comment by WilkAndy** [comment]
I couldn't replicate this failure, but i'm taking a guess as to what was wrong

File: modules/tensor_mechanics/src/materials/FiniteStrainMultiPlasticity.C [hunk 1]
```diff
@@ -222,13 +236,22 @@ FiniteStrainMultiPlasticity::plasticStep(const RankTwoTensor & stress_old, RankT
...
-    return_successful = returnMap(stress_good, stress, intnl_good, intnl, plastic_strain_good, plastic_strain, E_ijkl, dep, yf, iter, step_size <= _max_stepsize_for_dumb);
+    return_successful = returnMap(stress_good, stress, intnl_good, intnl, plastic_strain_good, plastic_strain, E_ijkl, dep, yf, iter, step_size <= _max_stepsize_for_dumb, linesearch_needed, ld_encountered, constraints_added);
     iterations += iter;
 
     if (return_successful)
     {
       num_consecutive_successes += 1;
       time_simulated += step_size;
+
+      //Moose::out << " returned_with_yf(0=neg,1=zero) ";
```
Comment (tonkmr) [inline modules/tensor_mechanics/src/materials/FiniteStrainMultiPlasticity.C]: Is there a reason you are keeping this commented out code?

File: modules/tensor_mechanics/src/materials/FiniteStrainMultiPlasticity.C [hunk 2]
```diff
@@ -339,7 +364,11 @@ FiniteStrainMultiPlasticity::returnMap(const RankTwoTensor & stress_old, RankTwo
   // those that exceed their _f_tol
   // active constraints.
   std::vector<bool> act;
-  buildActiveConstraints(f, stress, intnl, act);
+  buildActiveConstraints(f, stress, intnl, E_ijkl, act);
+
+  //Moose::out << "initial_act= ";
+  //for (unsigned i = 0 ; i < act.size() ; ++i)
```
Comment (tonkmr) [inline modules/tensor_mechanics/src/materials/FiniteStrainMultiPlasticity.C]: More commented out code

**Comment by tonkmr** [comment]
It looks like you kept in commented out debugging statements. It would be a lot cleaner if you could remove it.

**Comment by jwpeterson** [comment]
Feel free to disregard my comment if it's wrong, but the "rock" and "joint" stuff feels a bit more application-level to me while the tensor_mechanics module is meant to be more general?

**Comment by WilkAndy** [comment]
@tonkmr, yes, i can remove the commented Moose::out lines

a

**Comment by WilkAndy** [comment]
@jwpeterson , yes, those "rock" and "joint" stuff will really only be used in geomechanics applications, i think.  I'm not really where else to put them, however, and i would think that many users of TensorMechanics will want them (certainly all CSIRO people, and Rob Podgorney's team).   You could say exactly the same about "J2" plasticity - it's only really used in simulations involving metals, but it is very useful to have it in TensorMechanics.

I'm very happy to reorganise things - do you or @tonkmr have another suggestion?

a

**Comment by WilkAndy** [comment]
Removed those stupid Moose::out lines, thanks @tonkmr ,

a

**Comment by tonkmr** [comment]
I've had the same discussion about how "general" things need to be to be in a module with Derek, and the conclusion was that as long as it is not specific to one material (say, copper), then it should stay in the module. Every constitutive model is really only applicable to some applications, including linear elasticity.

**Comment by WilkAndy** [comment]
OK, that's cool, thanks @tonkmr.  Currently 9:57pm in Idaho Falls!

a

**Comment by WilkAndy** [comment]
Thanks @tonkmr ,

a

**Pull request merged**